File: conformance/m19/go/main.go
package main

import (
	"context"
	"fmt"
	"os"

	"github.com/apache/arrow-adbc/go/adbc"
	"github.com/apache/arrow-adbc/go/adbc/driver/flightsql"
	"github.com/apache/arrow-go/v18/arrow/memory"
)

func required(name string) string {
	value := os.Getenv(name)
	if value == "" {
		panic(name + " is required")
	}
	return value
}

func main() {
	ctx := context.Background()
	uri := os.Getenv("QCLI_FLIGHT_URI")
	if uri == "" {
		uri = "grpc://127.0.0.1:32010"
	}
	target := os.Getenv("QCLI_FLIGHT_TARGET")
	if target == "" {
		target = "demo"
	}
	query := os.Getenv("QCLI_FLIGHT_QUERY")
	if query == "" {
		query = "select * from sample"
	}
	expectedRows := int64(2)
	if os.Getenv("QCLI_FLIGHT_EXPECTED_ROWS") == "1" {
		expectedRows = 1
	}
	driver := flightsql.NewDriver(memory.DefaultAllocator)
	database, err := driver.NewDatabase(map[string]string{
		adbc.OptionKeyURI:                                   uri,
		flightsql.OptionAuthorizationHeader:                 "Bearer " + required("QCLI_FLIGHT_TOKEN"),
		flightsql.OptionRPCCallHeaderPrefix + "qcli-target": target,
		flightsql.OptionTimeoutQuery:                        "10",
		flightsql.OptionTimeoutFetch:                        "10",
	})
	if err != nil {
		panic(err)
	}
	defer database.Close()
	connection, err := database.Open(ctx)
	if err != nil {
		panic(err)
	}
	defer connection.Close()
	statement, err := connection.NewStatement()
	if err != nil {
		panic(err)
	}
	defer statement.Close()
	if err := statement.SetSqlQuery(query); err != nil {
		panic(err)
	}
	reader, _, err := statement.ExecuteQuery(ctx)
	if err != nil {
		panic(err)
	}
	rows := int64(0)
	for reader.Next() {
		rows += reader.RecordBatch().NumRows()
	}
	if err := reader.Err(); err != nil {
		panic(err)
	}
	reader.Release()
	if rows != expectedRows {
		panic(fmt.Sprintf("expected %d rows, got %d", expectedRows, rows))
	}
	objects, err := connection.GetObjects(ctx, adbc.ObjectDepthTables, nil, nil, nil, nil, nil)
	if err != nil {
		panic(err)
	}
	if !objects.Next() || objects.RecordBatch().NumRows() == 0 {
		panic("metadata returned no catalogs")
	}
	objects.Release()
	fmt.Println("go-adbc: PASS")
}

File: conformance/m19/rust/src/main.rs
use adbc_core::options::{AdbcVersion, OptionDatabase};
use adbc_core::{Connection, Database, Driver, Statement};
use adbc_driver_flightsql::DRIVER_PATH;
use adbc_driver_manager::ManagedDriver;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let uri = std::env::var("QCLI_FLIGHT_URI").unwrap_or_else(|_| "grpc://127.0.0.1:32010".into());
    let token = std::env::var("QCLI_FLIGHT_TOKEN")?;
    let target = std::env::var("QCLI_FLIGHT_TARGET").unwrap_or_else(|_| "demo".into());
    let query =
        std::env::var("QCLI_FLIGHT_QUERY").unwrap_or_else(|_| "select * from sample".into());
    let expected_rows = std::env::var("QCLI_FLIGHT_EXPECTED_ROWS")
        .unwrap_or_else(|_| "2".into())
        .parse::<usize>()?;
    let mut driver =
        ManagedDriver::load_dynamic_from_filename(DRIVER_PATH, None, AdbcVersion::default())?;
    let database = driver.new_database_with_opts([
        (OptionDatabase::Uri, uri.into()),
        (
            OptionDatabase::Other("adbc.flight.sql.authorization_header".into()),
            format!("Bearer {token}").into(),
        ),
        (
            OptionDatabase::Other("adbc.flight.sql.rpc.call_header.qcli-target".into()),
            target.into(),
        ),
    ])?;
    let mut connection = database.new_connection()?;
    let mut statement = connection.new_statement()?;
    statement.set_sql_query(&query)?;
    let mut rows = 0;
    for batch in statement.execute()? {
        rows += batch?.num_rows();
    }
    if rows != expected_rows {
        return Err(format!("expected {expected_rows} rows, got {rows}").into());
    }
    println!("rust-c-driver-manager-adbc: PASS");
    Ok(())
}
